File: DemoServer/AdditionalLanguages/go/StaticIndexes/mapReduceIndex/MapReduceIndex.go
package main

//region Usings
import "github.com/ravendb/ravendb-go-client"
//endregion

var globalDocumentStore *ravendb.DocumentStore

func main() {
    createDocumentStore()
    createDatabase()
    mapReduceIndex("USA")
    globalDocumentStore.Close()
}

func createDocumentStore() (*ravendb.DocumentStore, error) {
    if globalDocumentStore != nil {
        return globalDocumentStore, nil
    }
    urls := []string{"http://localhost:8080"}
    store := ravendb.NewDocumentStore(urls, "testGO")
    err := store.Initialize()
    if err != nil {
        return nil, err
    }
    globalDocumentStore = store
    return globalDocumentStore, nil
}

func createDatabase() {
    databaseRecord := ravendb.NewDatabaseRecord()
    databaseRecord.DatabaseName = "testGO"
    createDatabaseOperation := ravendb.NewCreateDatabaseOperation(databaseRecord, 1)
    var err = globalDocumentStore.Maintenance().Server().Send(createDatabaseOperation)
    if err != nil {
        fmt.Printf("d.store.Maintenance().Server().Send(createDatabaseOperation) failed with %s\n", err)
    }
}

//region Demo
func mapReduceIndex(country string) error {

        //region Step_1
        indexName := "Employees/ByCountry"
        index := ravendb.NewIndexCreationTask(indexName)
        //endregion
        
        //region Step_2
        index.Map := `
              map('employees', function(e) {
                 return {
                     Country: e.Address.Country,
                     Count: 1
                 }
             })
        `
        //endregion
        //region Step_3
        index.Reduce := `
              groupBy(x => x.Country)
              .aggregate(g => {
                     return {
                         Country: g.key,
                         Count: g.values.reduce((count, val) => val.Count + count, 0)
                     }
              })
        `
        //endregion
        
        //region Step_4
        err := index.Execute(globalDocumentStore, nil, "")
        if err != nil {
            return err
        }
        //endregion
        
        session, err := globalDocumentStore.OpenSession("")
        if err != nil {
            return err
        }
        defer session.Close()
    
        //region Step_5
        query := session.QueryIndex(indexName)
        query = query.Where("Country", "==", country)
    
        var queryResult *struct {
            Country string
            Count   int
        }
    
        err = query.First(&queryResult)
        if err != nil {
            return err
        }
        //endregion
        
        return nil
    }
    
type Employee struct {
    ID          string
    LastName    string
    FirstName   string
    Title       string
    Address     *Address
    HiredAt     ravendb.Time
    HomePhone   string
    Extension   string
    ReportsTo   string
    Notes       []string
    Territories []string
}
type Address struct {
    Line1      string
    Line2      string
    City       string   
    Country    string
}    
//endregion

File: DemoServer/AdditionalLanguages/java/src/main/java/net/ravendb/demo/staticIndexes/mapReduceIndex/MapReduceIndex.java
package net.ravendb.demo.staticIndexes.mapReduceIndex;
//region Usings
import net.ravendb.client.documents.indexes.AbstractIndexCreationTask;
import net.ravendb.client.documents.session.IDocumentSession;
//endregion
import net.ravendb.demo.common.DocumentStoreHolder;

public class MapReduceIndex {

    //region Demo
    //region Step_1
    public static class Employees_ByCountry extends AbstractIndexCreationTask {
    //endregion

        public Employees_ByCountry() {
            //region Step_2
            map = "docs.Employees.Select(employee => new { " +
                "    country = employee.Address.Country, " +
                "    countryCount = 1 " +
                "})";
            //endregion

            //region Step_3
            reduce = "results.GroupBy(result => result.country).Select(g => new { " +
                "    country = g.Key, " +
                "    countryCount = Enumerable.Sum(g, x => x.countryCount) " +
                "})";
            //endregion
        }

        //region Step_4
        public static class Result {
            private String country;
            private int countryCount;

            public String getCountry() {
                return country;
            }

            public void setCountry(String country) {
                this.country = country;
            }

            public int getCountryCount() {
                return countryCount;
            }

            public void setCountryCount(int countryCount) {
                this.countryCount = countryCount;
            }
        }
        //endregion
    }
    //endregion

    public int run(RunParams runParams) {
        String country = runParams.getCountry();
        int numberOfEmployeesFromCountry;

        //region Demo
        try (IDocumentSession session = DocumentStoreHolder.store.openSession()) {
            //region Step_5
            Employees_ByCountry.Result queryResult = session.query(Employees_ByCountry.Result.class, Employees_ByCountry.class)
                .whereEquals("country", country)
                .firstOrDefault();

            numberOfEmployeesFromCountry = queryResult != null ? queryResult.getCountryCount() : 0;
            //endregion
        }
        //endregion

        return numberOfEmployeesFromCountry;
    }
    
    public static class RunParams {
        private String country;

        public String getCountry() {
            return country;
        }

        public void setCountry(String country) {
            this.country = country;
        }
    }
}
